app/auth/views/fido.py

```diff
@@ -0,0 +1,106 @@
+import json
+import secrets
+import webauthn
+from app.config import URL as SITE_URL
+from urllib.parse import urlparse
+
+from flask import request, render_template, redirect, url_for, flash, session
+from flask_login import login_user
+from flask_wtf import FlaskForm
+from wtforms import HiddenField, validators
+
+from app.auth.base import auth_bp
+from app.config import MFA_USER_ID
+from app.log import LOG
+from app.models import User
+from app.extensions import db
+
+
+class FidoTokenForm(FlaskForm):
+    sk_assertion = HiddenField("sk_assertion", validators=[validators.DataRequired()])
+
+
+@auth_bp.route("/fido", methods=["GET", "POST"])
+def fido():
+    # passed from login page
+    user_id = session.get(MFA_USER_ID)
+
+    # user access this page directly without passing by login page
+    if not user_id:
+        flash("Unknown error, redirect back to main page", "warning")
+        return redirect(url_for("auth.login"))
+
+    user = User.get(user_id)
+
+    if not (user and (user.fido_uuid is not None)):
+        flash("Only user with security key linked should go to this page", "warning")
+        return redirect(url_for("auth.login"))
+
+    fido_token_form = FidoTokenForm()
+
+    next_url = request.args.get("next")
+
+    rp_id = urlparse(SITE_URL).hostname
+
+    webauthn_user = webauthn.WebAuthnUser(
+            user.fido_uuid, user.email, user.name, False,
+            user.fido_credential_id, user.fido_pk, user.fido_sign_count, rp_id)
+
+    # Handling POST requests
+    if fido_token_form.validate_on_submit():
+        try:
+            sk_assertion = json.loads(fido_token_form.sk_assertion.data)
+        except Exception as e:
+            flash('Key verification failed. Error: Invalid Payload', "warning")
+            return redirect(url_for("auth.login"))
+                    
+        challenge = session['fido_challenge']
+        credential_id = sk_assertion['id']
+
+        webauthn_assertion_response = webauthn.WebAuthnAssertionResponse(
+            webauthn_user,
+            sk_assertion,
+            challenge,
+            SITE_URL,
+            uv_required=False
+        )
+
+        is_webauthn_verified = False
+        try:
+            new_sign_count = webauthn_assertion_response.verify()
+            is_webauthn_verified = True
+        except Exception as e:
+            flash('Key verification failed. Error: {}'.format(e), "warning")
```

Comment:
Does this return the full error to the user? In this case, maybe it's more secured and user-friendly to do a `LOG.error()` here and returns only user-friendly user error message.

Reply:
0052dad13ecb1e4cfe28034ef42e22aa019a6150